src/libcore/fmt/num.rs

```diff
@@ -259,6 +259,111 @@ macro_rules! impl_Display {
     };
 }
 
+macro_rules! impl_Exp {
+    ($($t:ident),* as $u:ident via $conv_fn:ident named $name:ident) => {
+        fn $name(
+            mut n: $u,
+            is_nonnegative: bool,
+            upper: bool,
+            f: &mut fmt::Formatter<'_>
+        ) -> fmt::Result {
+            // 39 digits (worst case u128) + 2 digits for exponent + E + . = 43
+            let mut buf = [MaybeUninit::<u8>::uninit(); 43];
+            let mut curr = buf.len() as isize - 3; //index for buf
+            let mut exponent = 0;
+            let buf_ptr = MaybeUninit::first_ptr_mut(&mut buf);
+            let lut_ptr = DEC_DIGITS_LUT.as_ptr();
+
+            // need at least 16 bits for the 4-characters-at-a-time to work.
+            assert!(crate::mem::size_of::<$u>() >= 2);
+
+            //count and remove trailing zeroes
+            while n >= 10 && n % 10 == 0 {
+                n /= 10;
+                exponent += 1;
+            }
+
+            // decode 2 more chars, if > 2 chars
+            while n >= 100 {
+                let d1 = ((n % 100) as isize) << 1;
+                n /= 100;
+                exponent += 2;
+                curr -= 2;
+                unsafe {
+                    ptr::copy_nonoverlapping(lut_ptr.offset(d1), buf_ptr.offset(curr), 2);
+                }
+            }
+
+            // n here is <= 99, so at most 2 chars long
+            let mut n = n as isize; // possibly reduce 64bit math
+
+            // decode second-to-last character if applicable
+            if n >= 10 {
+                curr -= 1;
+                unsafe {*buf_ptr.offset(curr) = (n as u8 % 10_u8) + b'0';}
+                n /= 10;
+                exponent += 1;
+            }
+            curr -= 1;
+            unsafe {
+                *buf_ptr.offset(curr) = b'.';
+                *buf_ptr.offset(buf.len() as isize - 3) = if upper {b'E'} else {b'e'};
+            }
+            curr -= 1;
+            unsafe {*buf_ptr.offset(curr) = (n as u8) + b'0';}
+            if exponent < 10 {
+                unsafe {*buf_ptr.offset(buf.len() as isize - 2) = (exponent as u8) + b'0';}
+            } else {
+                let off = (exponent % 100) << 1;
+                unsafe {
+                    ptr::copy_nonoverlapping(
+                        lut_ptr.offset(off),
+                        buf_ptr.offset(buf.len() as isize - 2),
+                        2
+                    );
+                }
+            }
+
+            let buf_slice = unsafe {
+                str::from_utf8_unchecked(
+                    slice::from_raw_parts(buf_ptr.offset(curr), buf.len() - curr as usize))
+            };
+            f.pad_integral(is_nonnegative, "", buf_slice)
+        }
+
+        $(
+            #[unstable(feature = "IntegerExpFormat", issue = "39479", reason = "untested")]
+            impl fmt::LowerExp for $t {
```

Comment:
`impl` is insta-stable, i.e. the `#[unstable]` here has no effect and you can use `format!("{:e}", 123)` without `#![feature(IntegerExpFormat)]` anywhere (BTW feature names should be written like `integer_exp_format`). So this PR will additionally require FCP from @rust-lang/libs before merging.

Reply:
What should I put here instead?

Reply:
@maxbla `#[stable(feature = "integer_exp_format", since = "1.42.0")]`